pyccel/parser/semantic.py

```diff
@@ -1615,7 +1615,7 @@ def _ensure_inferred_type_matches_existing(self, class_type, d_var, var, is_auga
             # to remove memory leaks
             new_expressions.append(Deallocate(var))
 
-        elif class_type != var.class_type:
+        elif class_type != var.class_type and not isinstance(class_type.datatype, GenericType):
```

Comment:
Please can you explain this change. I don't see it mentioned in the PR description

Reply:
this is now deprecated. I've used `__eq__()` of the `GenericType` class from your PR, and added new ones for `HomogeneousListType` and `HomogeneousSetType` 
38d214c4ea66a9a06b29bb66979c9f8ac116a853

Reply:
the trailing new lines error were fixed